src/dll/Hooks/ValidateScripts.cpp

```diff
@@ -89,3 +89,58 @@ bool Hooks::ValidateScripts::Detach()
     isAttached = result != NO_ERROR;
     return !isAttached;
 }
+
+std::string WritePopupMessage(const std::vector<ValidationError>& validationErrors,
+                              const std::vector<PluginSystem::PluginName>& incompatiblePlugins)
+{
+    std::unordered_set<std::string_view> faultyScriptFiles;
+
+    for (const auto& error : validationErrors)
+    {
+        auto ref = error.GetSourceRef();
+        if (ref)
+        {
+            faultyScriptFiles.insert(ref->file);
+        }
+    }
+
+    if (incompatiblePlugins.empty() && faultyScriptFiles.empty())
+    {
+        return {};
+    }
+
+    std::string message;
+
+    if (!incompatiblePlugins.empty())
+    {
+        fmt::format_to(std::back_inserter(message),
+                       "The following red4ext plugins could not be loaded because they are "
+                       "incompatible with the current version of the game:\n");
+
+        for (const auto& plugin : incompatiblePlugins)
+        {
+            fmt::format_to(std::back_inserter(message), L"- {}\n", plugin);
```

Comment:
funny this compiles, believe it had issues in CET at least before... mixing of wstring + string in fmt fails i believe, at least with default fmt settings...

if you will be swapping to wstring, would add the L prefix to all string literals here probably.

Reply:
it seems to work like this but I can change it

Reply:
would be nice at least from "hidden" cost of conversions that have to ve done on the background.

but as it compiles and there shouldnt really be anything truly wrong with this, id say do as you want, is more nice to have... is not like this would be called 1000x per frame 😁